src/charm.py

```diff
@@ -43,55 +46,94 @@ def new_func(self, *args, **kwargs):
             func(self, *args, **kwargs)
         except BlockedStatusError as e:
             self.unit.status = BlockedStatus(str(e))
+
     return new_func
 
 
 class AlertmanagerCharm(CharmBase):
-    _stored = StoredState()
-
     def __init__(self, *args):
-        log.debug('Initializing charm.')
+        log.debug("Initializing charm.")
         super().__init__(*args)
         self.framework.observe(self.on.config_changed, self.on_config_changed)
-        self.framework.observe(self.on['alerting'].relation_changed, self.on_alerting_changed)
+        self.framework.observe(
+            self.on["alerting"].relation_changed, self.on_alerting_changed
+        )
+        self.framework.observe(
+            self.on["alertmanager"].relation_changed, self.on_alertmanager_changed
+        )
+        self.framework.observe(
+            self.on["alertmanager"].relation_departed, self.on_alertmanager_departed
+        )
 
     @status_catcher
     def on_alerting_changed(self, event):
+        self.update_alerting(event.relation)
```

Comment:
Doesn't `on_alerting_changed` seem redundant if all you are doing is making a method call to another method ? Or would it be preferable to separate this workflow into two methods with distinct responsibilites
1. One that checks if using is leader and requests the other method to generate the list of alerting addresses. And then sets the so obtained relation data.
2. Another that only generates the alerting relation data.

Wouldn't it be preferable to put the `is_leader` check in this event relation handler, thereby making it immediately obvious that the event is handled only by the leader ? (and a similar comment for other such cases in this pull request).

Reply:
I'm not sure I understand the comment here. update_alerting is used in multiple places and I don't want class code calling a callback.

See above comment about leader checking.